File: Pod/Classes/UIViewController+LCC.m
//
//  UIViewController+LCC.m
//  Pods
//
//  Created by Yuki Takahashi on 2015/10/13.
//
//

#import "UIViewController+LCC.h"
#import "ISMethodSwizzling.h"
#import "NSObject+LCC.h"

@implementation UIViewController (LCC)

+ (void)load
{
    @autoreleasepool {
        ISSwizzleInstanceMethod([self class],
                                @selector(viewDidLayoutSubviews),
                                @selector(lcc_viewDidLayoutSubviews));
        ISSwizzleInstanceMethod([self class],
                                @selector(viewDidAppear:),
                                @selector(lcc_viewDidAppear:));
    }
}

#pragma mark - Life cycle

- (void)lcc_viewDidLayoutSubviews
{
    [self lcc_viewDidLayoutSubviews];
    [self setLcc_isViewLaidOut:YES];
}

- (void)lcc_viewDidAppear:(BOOL)animated
{
    [self lcc_viewDidAppear:animated];
    [self setLcc_isViewAppeared:YES];
}

#pragma mark - flgs

- (BOOL)lcc_isViewLaidOut
{
    return [self lcc_flgForKey:@selector(lcc_isViewLaidOut)];
}

- (void)setLcc_isViewLaidOut:(BOOL)lcc_isViewLaidOut
{
    [self lcc_setFlg:lcc_isViewLaidOut forKey:@selector(lcc_isViewLaidOut)];
}

- (BOOL)lcc_isViewAppeared
{
    return [self lcc_flgForKey:@selector(lcc_isViewAppeared)];
}

- (void)setLcc_isViewAppeared:(BOOL)lcc_isViewAppeared
{
    [self lcc_setFlg:lcc_isViewAppeared forKey:@selector(lcc_isViewAppeared)];
}

@end

File: Pod/Classes/UIViewController+LCC.swift
//
//  UIViewController+LCC.swift
//  Pods
//
//  Created by Yuki Takahashi on 2015/10/9.
//
//

import UIKit
import Aspects

// MARK: - viewDidLoad

public extension UIViewController {
    
    public func afterViewDidLoad(block: () -> ()) {
        try! afterViewDidLoadSafety(block)
    }
    
    public func afterViewDidLoadSafety(block: () -> ()) throws {
        if self.isViewLoaded() {
            block()
            return
        }
        try self.dynamicType.aspect_hookSelector(
            Selector("viewDidLoad"),
            withOptions: AspectOptions.PositionAfter.union(.OptionAutomaticRemoval),
            usingVoidBlock: { (_) in block() }
        )
    }
    
}

// MARK: - viewDidAppear

public extension UIViewController {
    
    public func afterViewDidAppear(block: () -> ()) {
        try! afterViewDidAppearSafety(block)
    }
    
    public func afterViewDidAppearSafety(block: () -> ()) throws {
        if self.lcc_isViewAppeared {
            block()
            return
        }
        try self.dynamicType.aspect_hookSelector(
            Selector("viewDidAppear:"),
            withOptions: AspectOptions.PositionAfter.union(.OptionAutomaticRemoval),
            usingVoidBlock: { (_) in
                var key = Selector("lcc_isViewAppeared")
                self.lcc_setFlg(true, forKey: &key)
                block()
            }
        )
    }
    
}